File: src/lib/heart-rate-zones.ts
/**
 * Heart-rate zone metadata — mirrors HeartRateZones.swift (Watch app
 * target) exactly. The *current* zone during a live cardio session is
 * always computed once, on the Watch (the only device with an actual HR
 * sensor), and pushed to the phone as a plain number — this file only
 * needs the static band definitions (percentages, bpm ranges, labels) to
 * render the zone band UI, never an independent live-zone calculation.
 *
 * Zones follow Apple's own Watch Heart Rate Zones feature: 5 zones as
 * percentage bands of max heart rate (Karvonen-free, the simple
 * 220-minus-age estimate — no resting-HR calibration on file for this
 * user yet).
 */

export const MAX_HEART_RATE = 199; // 220 - 21

export type HeartRateZone = 1 | 2 | 3 | 4 | 5;

export type HeartRateZoneBand = {
  zone: HeartRateZone;
  minPct: number;
  maxPct: number;
  labelDe: string;
};

export const HEART_RATE_ZONE_BANDS: HeartRateZoneBand[] = [
  { zone: 1, minPct: 0.5, maxPct: 0.6, labelDe: "Erholung" },
  { zone: 2, minPct: 0.6, maxPct: 0.7, labelDe: "Fettverbrennung" },
  { zone: 3, minPct: 0.7, maxPct: 0.8, labelDe: "Ausdauer" },
  { zone: 4, minPct: 0.8, maxPct: 0.9, labelDe: "Schwelle" },
  { zone: 5, minPct: 0.9, maxPct: 1.5, labelDe: "Maximal" },
];

export function heartRateZoneBpmRange(zone: HeartRateZone): { min: number; max: number } {
  const band = HEART_RATE_ZONE_BANDS.find((b) => b.zone === zone)!;
  return {
    min: Math.round(band.minPct * MAX_HEART_RATE),
    max: Math.round(Math.min(band.maxPct, 1) * MAX_HEART_RATE),
  };
}

export function heartRateZoneLabel(zone: HeartRateZone): string {
  return HEART_RATE_ZONE_BANDS.find((b) => b.zone === zone)!.labelDe;
}

File: ios/App/FitTrackWatch Watch App/HeartRateZones.swift
//
//  HeartRateZones.swift
//  FitTrackWatch Watch App
//
//  Mirrors src/lib/heart-rate-zones.ts exactly. This is the only place the
//  *current* zone is ever computed — the Watch is the only device with an
//  actual HR sensor, so it determines the zone once here and pushes the
//  resulting number to the phone (see PhoneWorkoutObserver.pushCardioLiveUpdate)
//  rather than the phone re-deriving it from a relayed bpm value, which
//  would risk two independently-implemented formulas drifting apart.
//

import Foundation

enum HeartRateZones {
    /// 220 - 21 (see src/lib/heart-rate-zones.ts for the same constant).
    static let maxHeartRate: Double = 199

    /// (zone, minPct, maxPct) — percentage bands of maxHeartRate, matching
    /// Apple's own Watch Heart Rate Zones feature.
    private static let bands: [(zone: Int, minPct: Double, maxPct: Double)] = [
        (1, 0.5, 0.6),
        (2, 0.6, 0.7),
        (3, 0.7, 0.8),
        (4, 0.8, 0.9),
        (5, 0.9, 1.5),
    ]

    /// nil below Zone 1's lower bound (resting/warming up — not yet in any
    /// zone) or for a non-positive/missing reading.
    static func zone(forBpm bpm: Double) -> Int? {
        guard bpm > 0 else { return nil }
        let pct = bpm / maxHeartRate
        guard pct >= bands[0].minPct else { return nil }
        for band in bands where pct >= band.minPct && pct < band.maxPct {
            return band.zone
        }
        return 5
    }

    /// bpm bounds for a zone — drives the position pointer within the zone
    /// band (see ZoneIndicatorView), i.e. "just entered Zone 3" vs. "about
    /// to cross into Zone 4," not just which zone.
    static func bpmRange(forZone zone: Int) -> (min: Double, max: Double) {
        let band = bands.first { $0.zone == zone } ?? bands[bands.count - 1]
        return (band.minPct * maxHeartRate, min(band.maxPct, 1) * maxHeartRate)
    }

    /// 0 at the zone's lower bound, 1 at its upper bound, clamped — where
    /// exactly `bpm` sits within `zone`'s range right now.
    static func positionWithinZone(bpm: Double, zone: Int) -> Double {
        let range = bpmRange(forZone: zone)
        guard range.max > range.min else { return 0 }
        return min(1, max(0, (bpm - range.min) / (range.max - range.min)))
    }

    static func labelDe(forZone zone: Int) -> String {
        switch zone {
        case 1: return "Erholung"
        case 2: return "Fettverbrennung"
        case 3: return "Ausdauer"
        case 4: return "Schwelle"
        default: return "Maximal"
        }
    }
}
